### PR TITLE
Add batch and epoch callbacks to the trainer

### LINKED ISSUE #3218: Predict during training
**Is your feature request related to a problem? Please describe.**
I would like to log model predictions for every epoch, even though I am only keeping checkpoints of the latest 3 epochs. This enables post-analysis on the predictions without the disk overhead of logging model weights every epoch. One use case is to generate training curves for new metrics on old (expensive) training runs.

**Describe the solution you'd like**
Either a Predict callback or a "predict" field added to the Validate callback

I looked into subclassing or modifying the Validate callback, but since I need the `output_dict`, I would also need to alter `CallbackTrainer.batch_loss()` to return the `output_dict` instead of just returning loss. For now, I have written a hack that duplicates CallbackTrainer.batch_loss() functionality inside the callback.

### LINKED ISSUE #3913: Make the trainer more configurable
From https://github.com/allenai/allennlp/issues/3519#issuecomment-588419410:

Ok, I'm scoping this out right now, taking the approach of adding a few targeted callbacks to the original trainer, and simplifying what's there / making it more modular.  Here's a list of specific items that I'm thinking of doing; I'm putting this here to get feedback before starting on this.  My process for deciding on these things to change was (1) see what parameters to `__init__` I can remove / push to other, configurable dependencies, (2) see what other places might want some customization, and propose specific callbacks for them.

1. ~Change the name of `TrainerBase` to `Trainer`, and `Trainer` to `GradientDescentTrainer`.  This is just a cosmetic change that makes the trainer code consistent with other naming conventions in the library, and makes it so we can use `Trainer` annotations in places where it's appropriate, instead of the more awkward-looking `TrainerBase`.  I could maybe be persuaded to just make it one class, even, that's still `Registrable` and can be subclassed (like `Embedding` is).~

2. ~Make `TensorboardWriter` an argument to `__init__`, and to `from_partial_objects` (with a `Lazy` annotation), so I can remove several of the `__init__` parameters that are specific to the tensorboard writer.  Specifically, that's these parameters: https://github.com/allenai/allennlp/blob/dd4ee62cf55e5fe50c870152501f909203a6b1d7/allennlp/training/trainer.py#L59-L62~

3. ~Move the tensorboard logging logic into the `TensorboardWriter` class, so it's more easily configurable with your own `TensorboardWriter`:~ https://github.com/allenai/allennlp/blob/dd4ee62cf55e5fe50c870152501f909203a6b1d7/allennlp/training/trainer.py#L445-L464 ~This also could let us remove this parameter:~ https://github.com/allenai/allennlp/blob/dd4ee62cf55e5fe50c870152501f909203a6b1d7/allennlp/training/trainer.py#L63 ~Question here: I could generalize `TensorboardWriter` to handle other kinds of logging, and just add a bunch of methods / calls to it, which might simplify some things.  If the only reason we want callbacks is to provide better logging, we could accomplish that with a specific, overridable `TrainLogger` class, or something, which is a generalization of the `TensorboardWriter`.  Does this would probably also let me completely remove the `serialization_dir` parameter to the `Trainer`.  It's tempting to also rip out tqdm and roll our own thing somehow (or figure out a way to push it into the `TrainLogger` or whatever it gets called), as that would simplify the trainer and fix some bugs~.  Decision on the question: add a couple of simple callbacks, one for end of each batch, and one for end of each epoch.  See item 7.

4. ~Remove these two parameters, as they are redundant with the `Checkpointer` (and were already removed from `from_partial_objects`):~ https://github.com/allenai/allennlp/blob/dd4ee62cf55e5fe50c870152501f909203a6b1d7/allennlp/training/trainer.py#L50-L51  ~I'll make sure that the default `Checkpointer` behavior is reasonable if none is passed.~  Also, move this parameter into the `Checkpointer` and remove it from the `Trainer`: https://github.com/allenai/allennlp/blob/dd4ee62cf55e5fe50c870152501f909203a6b1d7/allennlp/training/trainer.py#L53

5. ~On `grad_norm` and `grad_clipping`: I could try to remove these, but I'm not sure it's worth it.  This would mean adding some kind of callback, except it's a weird callback, because clipping requires setting hooks on the model before starting training.  I'm leaning towards just leaving these two alone; they don't add much complexity to the `Trainer`, and we don't get issues asking to configure this more than what we already allow.~ (Nothing to do, crossing it off the list)

6. ~I don't think `num_gradient_accumulation_steps` can be simplified or put into a callback.  There are three parameters for distributed training, which I could *maybe* simplify to one object, but that doesn't really seem worth it.  All of the rest of the parameters are either core training parameters or configurable objects that don't seem like they need to be further separated out.~ (Nothing to do, crossing it off the list)

7. Add two very simple callbacks, one that gets called at the end of each batch (getting passed the batch inputs and outputs, and the trainer object), and one that gets called at the end of each epoch (getting metrics and the trainer object).  This allows for an incredible amount of flexibility, for logging, saving predictions, or whatever you want.

### PR DESCRIPTION
**Pull request by matt-gardner**
Fixes #3913 and fixes #3218.  This is NOT a fix for #3785, which I thought might be covered in the work for #3913, but it turns out was unrelated.

### CODE REVIEW

File: allennlp/training/trainer.py [hunk 1]
```diff
@@ -126,6 +160,8 @@ def __init__(
         momentum_scheduler: Optional[MomentumScheduler] = None,
         tensorboard_writer: TensorboardWriter = None,
         moving_average: Optional[MovingAverage] = None,
+        batch_callback: Optional[BatchCallback] = None,
+        epoch_callback: Optional[EpochCallback] = None,
```
Comment (dirkgr) [inline allennlp/training/trainer.py]: You could make these `List`s instead of `Optional`, and call all of them. If we're going the callback route, we might as well go all in.

Comment (matt-gardner) [reply allennlp/training/trainer.py]: Note that we are not providing any implementation of a callback here.  Anything a person wants to do will be code they write.  If they want to have several things done, they can just put those things in a callback.  If they want to have composable things in their callbacks, they have modularize it however they want (including a callback that calls a list of functions).

Comment (dirkgr) [reply allennlp/training/trainer.py]: I can imagine a few orthogonal, default callbacks we could provide. Can we implement `TensorboardWriter` as one of these? Writing predictions to a file? Automatic logging to W&B?

The advantage of using callbacks is configurability and composability. It would be a neat feature if you can just add what you need to this list of callbacks.

Comment (matt-gardner) [reply allennlp/training/trainer.py]: Fair enough, you make a good point.  I don't think we can pull the tensorboard writer out into a callback, because of how computing gradient updates works, but you're right that you could imagine some built-in ones that get added later, and being able to pass in a list makes sense.  It was easy enough to change.

**Comment by mahnerak** [comment]
Is there any reason why are there `BatchCallback` and `EpochCallback` instead of a single `Callback` class that could handle both? (+a callback at the end of the training, as it is done in Keras, PyTorch-Lightning, and was done for CallbackTrainer)

Say, I have a custom logger which tracks both epoch and batch. What's the recommended way to share an open connection/file (and when to close it when finished)?

**Comment by dirkgr** [comment]
This is an interesting scenario. As a workaround, you can stick stuff into the trainer that you want to share between the two. It's not super elegant, but it will work.

`BatchCallback` and `EpochCallback` receive different parameters, so that's why they are separate. But I see your point. Maybe in version 2.0 we will add a `TrainerCallback` that can do both, to address this scenario better.
